src/robot_localization_estimator.cpp

```diff
@@ -0,0 +1,330 @@
+
+  yCoeff = cp * cr;
+  zCoeff = -cp * sr;
+  double dFz_dR = (yCoeff * yVel + zCoeff * zVel) * delta +
+                  (yCoeff * yAcc + zCoeff * zAcc) * oneHalfATSquared;
+  double dFY_dR = (yCoeff * pitchVel + zCoeff * yawVel) * delta;
+
+  xCoeff = -cp;
+  yCoeff = -sp * sr;
+  zCoeff = -sp * cr;
+  double dFz_dP = (xCoeff * xVel + yCoeff * yVel + zCoeff * zVel) * delta +
+                  (xCoeff * xAcc + yCoeff * yAcc + zCoeff * zAcc) * oneHalfATSquared;
+  double dFY_dP = (xCoeff * rollVel + yCoeff * pitchVel + zCoeff * yawVel) * delta;
+
+  // Much of the transfer function Jacobian is identical to the transfer function
+  Eigen::MatrixXd transferFunctionJacobian = transfer_function;
+  transferFunctionJacobian(StateMemberX, StateMemberRoll) = dFx_dR;
+  transferFunctionJacobian(StateMemberX, StateMemberPitch) = dFx_dP;
+  transferFunctionJacobian(StateMemberX, StateMemberYaw) = dFx_dY;
+  transferFunctionJacobian(StateMemberY, StateMemberRoll) = dFy_dR;
+  transferFunctionJacobian(StateMemberY, StateMemberPitch) = dFy_dP;
+  transferFunctionJacobian(StateMemberY, StateMemberYaw) = dFy_dY;
+  transferFunctionJacobian(StateMemberZ, StateMemberRoll) = dFz_dR;
+  transferFunctionJacobian(StateMemberZ, StateMemberPitch) = dFz_dP;
+  transferFunctionJacobian(StateMemberRoll, StateMemberRoll) = dFR_dR;
+  transferFunctionJacobian(StateMemberRoll, StateMemberPitch) = dFR_dP;
+  transferFunctionJacobian(StateMemberRoll, StateMemberYaw) = dFR_dY;
+  transferFunctionJacobian(StateMemberPitch, StateMemberRoll) = dFP_dR;
+  transferFunctionJacobian(StateMemberPitch, StateMemberPitch) = dFP_dP;
+  transferFunctionJacobian(StateMemberPitch, StateMemberYaw) = dFP_dY;
+  transferFunctionJacobian(StateMemberYaw, StateMemberRoll) = dFY_dR;
+  transferFunctionJacobian(StateMemberYaw, StateMemberPitch) = dFY_dP;
+
+  // Project the state forward: x = Ax (really, x = f(x))
+  state_at_req_time.state = transfer_function * boundary_state.state;
+
+  // Handle wrapping
+  wrapStateAngles(state_at_req_time);
+
+  // Project the error forward: P = J * P * J' + Q
+  state_at_req_time.covariance = (transferFunctionJacobian * boundary_state.covariance *
+                                  transferFunctionJacobian.transpose());
+  state_at_req_time.covariance.noalias() += (processNoiseCovariance_ * delta);
+
+  return;
+}
+
+void RobotLocalizationEstimator::interpolate(const EstimatorState& given_state_1,
+                                             const EstimatorState& given_state_2,
+                                             const double requested_time,
+                                             EstimatorState& state_at_req_time) const
+{
+  /*
+   * TODO: Right now, we only extrapolate from the last known state before the requested time.
+   * But as the state after the requested time is also known, we may want to perform
+   * interpolation between states.
+   */
+  extrapolate(given_state_1, requested_time, state_at_req_time);
```

Comment:
I found a method for blending the covariance matrix [here](http://www.agi.com/downloads/resources/white-papers/CovarianceInterpolationFinal.pdf). But so far I haven't been able to find a good way for blending/interpolation of the Euler angles in the state representation, taking into account the wrapped state angles and velocity constriants on both sides. Do you have suggestions for this?